### PR TITLE
Check performance improvements

### PR DESCRIPTION
**Pull request by LucGenetier**
Expression `Power(x.y,3) <  100 ` performance improvement of 38%

- replaces IEnumerable<TexlFunction> with TexlFunctionSet<T> which is a Dictionary<string, HashSet<T>>
- T is used to either store a TexlFunction or TexlFunctionImplementation which adds a function pointer (AsyncFunctionPtr) in interpreter
- added cache in ComposedReadOnlySymbolsTable INameResolver.Functions
- added new test project Microsoft.PowerFx.Performance.Tests using BenchmarkDotNet

image
U = unoptimized, O = optimized, G = gain in %

### CODE REVIEW

File: src/Directory.Build.props [hunk 1]
```diff
@@ -58,13 +58,17 @@
     <Product>Microsoft Bot Builder SDK</Product>
     <Copyright>© Microsoft Corporation. All rights reserved.</Copyright>
   </PropertyGroup>
-
-  <PropertyGroup>
-    <AssemblyOriginatorKeyFile>$(SolutionDir)\build\35MSSharedLib1024.snk</AssemblyOriginatorKeyFile>
+  
+  <PropertyGroup Condition="'$(Project)'!='Microsoft.PowerFx.Performance.Tests'">
```
Comment (MikeStall) [inline src/Directory.Build.props:62]: >PropertyGroup [](http://example.com/codeflow?start=3&length=13)

Can you add a <! -- --> comment explaining this?

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs [hunk 2]
```diff
@@ -28,5 +28,10 @@ public enum Features : int
         /// Filter(A, A[@Value] = 2)
         /// </summary>
         DisableRowScopeDisambiguationSyntax = 0x4,
+
+        /// <summary>
+        /// All features enabled
+        /// </summary>
+        All = ~0
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs:35]: > All = ~0 [](http://example.com/codeflow?start=7&length=9)

separate PR

File: src/tests/Microsoft.PowerFx.Performance.Tests/PerformanceTest1.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            record = FormulaValue.NewRecordFromFields(new NamedValue("y", FormulaValue.New(11)));
+            symbolTable = new SymbolTable { DebugName = "PerformanceTest1" };
+            slot = symbolTable.AddVariable("x", record.Type);
+            engine = new Engine(new PowerFxConfig(new CultureInfo("en-US"), Features.All));
+        }
+
+        [Benchmark]
+        public CheckResult Check()
+        {
+            var expr = "Power(x.y,3) <  100 ";
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Performance.Tests/PerformanceTest1.cs:34]: >Power(x.y,3) <  100  [](http://example.com/codeflow?start=24&length=20)


1. Try with 5 function calls 
2. What's split between Lex/Parse/Bind

File: src/tests/Microsoft.PowerFx.Performance.Tests/PerformanceTest1.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+using System.Globalization;
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Diagnostics.Windows;
+using BenchmarkDotNet.Diagnostics.Windows.Configs;
+using Microsoft.PowerFx.Core.Tests;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Performance.Tests
+{
+    [EtwProfiler]
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Performance.Tests/PerformanceTest1.cs:13]: >[EtwProfiler] [](http://example.com/codeflow?start=4&length=13)

what does this mean? Why not just a timer.

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs [hunk 5]
```diff
@@ -60,18 +60,19 @@ public override FormulaType GetTypeFromSlot(ISymbolSlot slot)
...
+        TexlFunctionSet<TexlFunction> __functions = null;
+
         // Expose the list to aide in intellisense suggestions. 
-        IEnumerable<TexlFunction> INameResolver.Functions
+        TexlFunctionSet<TexlFunction> INameResolver.Functions
         {
             get
             {
-                foreach (INameResolver table in _symbolTables)
+                if (__functions == null)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs:70]: >__functions [](http://example.com/codeflow?start=20&length=11)

cache invalidation?

Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs:70]: Ie, rebase and pass this test https://github.com/microsoft/Power-Fx/pull/886

SymbolTables do have a VersionHash you can use for invalidation.

File: src/Directory.Build.props [hunk 6]
```diff
@@ -58,13 +58,17 @@
...
+    <AssemblyOriginatorKeyFile Condition="EXISTS('..\..\..\src\build\35MSSharedLib1024.snk')">..\..\..\src\build\35MSSharedLib1024.snk</AssemblyOriginatorKeyFile>
```
Comment (lesaltzm) [inline src/Directory.Build.props:63]: >Condition="EXISTS('..\..\..\src\build\35MSSharedLib1024.snk')" [](http://example.com/codeflow?start=31&length=62)

This is part of the repo, when doesn't it exist?

File: src/libraries/Microsoft.PowerFx.Core/AssemblyProperties.cs [hunk 7]
```diff
@@ -40,5 +40,6 @@
...
 [assembly: InternalsVisibleTo("Microsoft.PowerFx.Interpreter.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9")]
+[assembly: InternalsVisibleTo("Microsoft.PowerFx.Performance.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9")]
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/AssemblyProperties.cs:43]: >Microsoft.PowerFx.Performance.Tests [](http://example.com/codeflow?start=31&length=35)

If we aren't strongnaming the new project (assuming from change to Directory.Build.props), does this internalsvisibleto actually work?

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionImplementation.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    internal class TexlFunctionImplementation : ITexlFunction
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionImplementation.cs:8]: >TexlFunctionImplementation [](http://example.com/codeflow?start=19&length=26)

Comment about what this is?

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    internal class TexlFunctionSet<T>
+        where T : ITexlFunction
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:14]: >T : ITexlFunction [](http://example.com/codeflow?start=14&length=17)

Why do we need this interface/generic param? When will this not be the concrete `TexlFunction`?

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+        internal TexlFunctionSet(T function)
+            : this()
+        {
+            if (function == null)
+            {
+                throw new ArgumentNullException($"{nameof(function)} cannot be null", nameof(function));
+            }
+
+            _functions.Add(function.Name, new HashSet<T>() { function });
+            _functionsInvariant.Add(function.LocaleInvariantName, new HashSet<T>() { function });
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:36]: >LocaleInvariantName [](http://example.com/codeflow?start=45&length=19)

I think, although not 1000% sure, that LocaleInvariantName == Name and we never have locale-specific function names.

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs [hunk 11]
```diff
@@ -60,18 +60,19 @@ public override FormulaType GetTypeFromSlot(ISymbolSlot slot)
             return slot.Owner.GetTypeFromSlot(slot);
         }
 
+        TexlFunctionSet<TexlFunction> __functions = null;
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs:63]: >__functions [](http://example.com/codeflow?start=38&length=11)

nit: just one _ char prefix?

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs [hunk 12]
```diff
@@ -28,5 +28,10 @@ public enum Features : int
...
     }
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs:36]: Adding this seems potentially risky. We don't want hosts to use this and pick up future flags unintentionally, I'd think

File: src/Directory.Build.props (hunk 6 above)
Comment (LucGenetier) [reply src/Directory.Build.props:63]: BenchMarkNet creates a small project and builds it using this .props file and will not find the file as it's in a totally different folder

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionImplementation.cs (hunk 8 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionImplementation.cs:8]: Will add

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs (hunk 10 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:36]: I agree but wanted to be 100% compatibile even if this part is never used.
If LocalInvariantName is never used, let's remove it and we'll save CPU & memory

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+
+            foreach (var functionSet in functionSets)
+            {
+                Add(functionSet);
+            }
+        }
+
+        // Slow API, only use for backward compatibility
+        internal IEnumerable<T> Functions
+        {
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:84]: Good point - but a bit odd to make new code as [Obsolete] :)

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs (hunk 9 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:14]: We're in Core and can't access Interpreter with AsyncFunctionPtr

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs (hunk 11 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs:63]: I'll check but i think it conflicts

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs (hunk 5 above)
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs:70]: Yes. 
ReadOnly is like IReadOnlyDictionary - the underlying dictionary can still be changed.

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs (hunk 12 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs:36]: This is for testing where I want to consider them all, even future ones, I'll add a comment on the risk you raise

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs (hunk 2 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/Features.cs:35]: Done

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionImplementation.cs (hunk 8 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionImplementation.cs:8]: Also not placed in the right project, I moved it to Interpreter

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs (hunk 9 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:14]: Used in Interpreter, in TexlFunctionImplementation and Library.cs

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs (hunk 10 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:36]: Can be considered in a different PR if we need

**Comment by MikeStall** [comment]
Superceded by #1019

**Pull request closed without merging**